Makefile

```diff
@@ -0,0 +1,130 @@
+PHAN_PHP_VERSION=7.2
+export PHAN_DISABLE_XDEBUG_WARN=1
+
+# PHPStan
+PHPSTAN=vendor/bin/phpstan
+PHPSTAN_ARGS=analyse src tests --level=2 -c .phpstan.neon
+
+# Psalm
+PSALM=vendor/bin/psalm
+PSALM_ARGS=--show-info=false
+
+# Composer
+COMPOSER=$(PHP) $(shell which composer)
+
+# Infection
+INFECTION=vendor/bin/infection
+MIN_MSI=60
+MIN_COVERED_MSI=60
+INFECTION_ARGS=--min-msi=$(MIN_MSI) --min-covered-msi=$(MIN_COVERED_MSI) --threads=$(JOBS) --coverage=build/logs --log-verbosity=default --show-mutations
+
+all: test
+
+##############################################################
+# Continuous Integration                                     #
+##############################################################
+
+ci-test: SILENT=
+ci-test: prerequisites
+	$(SILENT) $(PHPDBG) $(PHPUNIT) $(PHPUNIT_COVERAGE_CLOVER) --group=$(PHPUNIT_GROUP)
+
+ci-analyze: SILENT=
+ci-analyze: prerequisites ci-phpunit ci-infection ci-phan ci-phpstan ci-psalm
+
+ci-phpunit: ci-cs
+	$(SILENT) $(PHPDBG) $(PHPUNIT) $(PHPUNIT_ARGS)
+	cp build/logs/junit.xml build/logs/phpunit.junit.xml
+
+ci-infection: ci-phpunit
+	$(SILENT) $(PHP) $(INFECTION) $(INFECTION_ARGS)
+
+ci-phan: ci-cs
+	$(SILENT) $(PHP) $(PHAN) $(PHAN_ARGS)
+
+ci-phpstan: ci-cs
+	$(SILENT) $(PHP) $(PHPSTAN) $(PHPSTAN_ARGS) --no-progress
+
+ci-psalm: ci-cs
+	$(SILENT) $(PHP) $(PSALM) $(PSALM_ARGS) --no-cache --shepherd
+
+ci-cs: prerequisites
+	$(SILENT) $(PHP) $(PHP_CS_FIXER) $(PHP_CS_FIXER_ARGS) --dry-run --stop-on-violation fix
+
+##############################################################
+# Development Workflow                                       #
+##############################################################
+
+test: phpunit analyze composer-validate
+
+.PHONY: composer-validate
+composer-validate: test-prerequisites
+	$(SILENT) $(COMPOSER) validate --strict
+
+test-prerequisites: prerequisites composer.lock
+
+phpunit: cs
+	$(SILENT) $(PHP) $(PHPUNIT) $(PHPUNIT_ARGS) --verbose
+	cp build/logs/junit.xml build/logs/phpunit.junit.xml
+	$(SILENT) $(PHP) $(INFECTION) $(INFECTION_ARGS)
+
+analyze: cs
+	$(SILENT) $(PHP) $(PHAN) $(PHAN_ARGS) --color
+	$(SILENT) $(PHP) $(PHPSTAN) $(PHPSTAN_ARGS)
+	$(SILENT) $(PHP) $(PSALM) $(PSALM_ARGS)
+
+cs: test-prerequisites
+	$(SILENT) $(PHP) $(PHP_CS_FIXER) $(PHP_CS_FIXER_ARGS) --diff fix
+
+##############################################################
+# Prerequisites Setup                                        #
+##############################################################
+
+# We need both vendor/autoload.php and composer.lock being up to date
+.PHONY: prerequisites
+prerequisites: report-php-version build/cache vendor/autoload.php .phan composer.lock infection.json.dist .phpstan.neon
+
+# Do install if there's no 'vendor'
+vendor/autoload.php:
+	$(SILENT) $(COMPOSER) install --prefer-dist
+	test -d vendor/infection/infection/src/StreamWrapper/ && rm -fr vendor/infection/infection/src/StreamWrapper/ && $(SILENT) $(COMPOSER) dump-autoload || true
```

Comment:
This is a workaround for testing with Infection, probably needs addressing later.

Reply:
could you please explain why is it needed?

Reply:
Yes, that's because we're installing Infection as a dev dependency with composer, and naturally we will have the same StreamWrapper classes here in the library, and there, inside Infection. We can remove this workaround once there's a new release of Infection that depends on this package.